src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2Client.cs

```diff
@@ -0,0 +1,440 @@
+        /// </summary>
+        /// <param name="authorizationCode">Authorization code.</param>
+        /// <returns>Token result.</returns>
+        Task<OAuth2TokenResult> GetTokenByAuthorizationCodeAsync(string authorizationCode);
+
+        /// <summary>
+        /// Use a refresh token to get a new access token.
+        /// </summary>
+        /// <param name="refreshToken">Refresh token.</param>
+        /// <returns>Token result.</returns>
+        Task<OAuth2TokenResult> GetTokenByRefreshTokenAsync(string refreshToken);
+
+        /// <summary>
+        /// Exchange a device code grant acquired from <see cref="GetDeviceCodeAsync"/> for an access token.
+        /// </summary>
+        /// <param name="deviceCodeResult">Device code grant result.</param>
+        /// <returns>Token result.</returns>
+        Task<OAuth2TokenResult> GetTokenByDeviceCodeAsync(OAuth2DeviceCodeResult deviceCodeResult);
+    }
+
+    public class OAuth2Client : IOAuth2Client
+    {
+        private readonly HttpClient _httpClient;
+        private readonly OAuth2ServerEndpoints _endpoints;
+        private readonly Uri _redirectUri;
+        private readonly string _clientId;
+        private readonly string _clientSecret;
+
+        public OAuth2Client(HttpClient httpClient, OAuth2ServerEndpoints endpoints, string clientId, Uri redirectUri = null, string clientSecret = null)
+        {
+            _httpClient = httpClient;
+            _endpoints = endpoints;
+            _clientId = clientId;
+            _redirectUri = redirectUri;
+            _clientSecret = clientSecret;
+        }
+
+        #region IOAuth2Client
+
+        public async Task<string> GetAuthorizationCodeAsync(IEnumerable<string> scopes, IWebBrowser browser)
+        {
+            if (_redirectUri is null)
+            {
+                throw new InvalidOperationException("A redirect URI is required for this flow but client was constructed without one.");
+            }
+
+            string state = GenerateRandomString();
+            string scopesStr = string.Join(" ", scopes);
+
+            var queryParams = new Dictionary<string, string>
+            {
+                [OAuth2Constants.AuthorizationEndpoint.ResponseTypeParameter] = OAuth2Constants.AuthorizationEndpoint.AuthorizationCodeResponseType,
+                [OAuth2Constants.ClientIdParameter] = _clientId,
+                [OAuth2Constants.AuthorizationEndpoint.StateParameter] = state,
+                [OAuth2Constants.RedirectUriParameter] = _redirectUri.ToString()
+            };
+
+            if (!string.IsNullOrWhiteSpace(scopesStr))
+            {
+                queryParams[OAuth2Constants.ScopeParameter] = scopesStr;
+            }
+
+            var requestUriBuilder = new UriBuilder(_endpoints.AuthorizationEndpoint)
+            {
+                Query = queryParams.ToQueryString()
+            };
+
+            Uri requestUri = requestUriBuilder.Uri;
+
+            // TODO: handle error response

```

Comment:
Error handling feels like something we should address before merging, even if it's just tracing the response code.